databuilder/query_engines/base_sql.py

```diff
@@ -493,153 +470,163 @@ def round_to_first_of_month(self, date):
-        if table in tables:
-            return query
-        join = sqlalchemy.join(
-            query.get_final_froms()[0],
-            table,
-            query.selected_columns.patient_id == table.c.patient_id,
-            isouter=True,
-        )
-        return query.select_from(join)
 
-    def generate_results_query(self):
-        """Query to generate the final single results table"""
-        # `population` is a special-cased boolean column, it doesn't appear
-        # itself in the output but it determines what rows are included
-        # Build the base results table from the population table
-        column_definitions = self.column_definitions.copy()
-        population = column_definitions.pop("population")
-        results_query = self.get_population_table_query(population)
-
-        # Build big JOIN query which selects the results
-        for column_name, output_node in column_definitions.items():
-            # For each output column, generate the query that selects it from its interim table(s)
-            # For most outputs there will just be a single interim table.  Category outputs
-            # may require more than one.
-            column, tables = self.get_value_expression(output_node)
-            # Then generate the query to join on it
-            for table in tables:
-                results_query = self.include_joined_table(results_query, table)
-
-            # Add this column to the final selected results
-            results_query = results_query.add_columns(column.label(column_name))
-
-        return results_query
-
-    def write_query_to_table(self, table, query):
-        """
-        Returns a new query which, when executed, writes the results of `query`
-        into `table`
-        """
-        raise NotImplementedError()
+def apply_optimisations(column_definitions):
+    """
+    Apply various transformations to the supplied query DAG which make it easier to
+    generate better performing SQL
+    """
+    # It's algorithmically easier to apply these transformations by modifying the graph
+    # in place, so we copy it first
+    column_definitions = copy.deepcopy(column_definitions)
+
+    reify_query_before_selecting_column(column_definitions)
+
+    return column_definitions
+
+
+def reify_query_before_selecting_column(column_definitions):
+    """
+    We sometimes need to be able to take a SQLAlchemy query and treat it as something
+    table-like from which we can select columns (a process we describe as
+    "reification"). In fact, SQLAlchemy will do this implictly for us: we can just
+    select columns from a query and it will automatically create a subquery for us.
+    However this triggers a warning not to rely on this behaviour. And in any case, we
+    want to control exactly how and when queries are reified. So here we walk the query
+    graph, find all the places where we select columns and inject an operation to
+    explicitly reify the query before doing so.
+    """
```

Comment:
I didn't understand this until I read the docstring for `get_element_from_reify_query`.  Could this docstring mention temporary tables?